case_studies/weak_lensing/lensing_encoder.py

```diff
@@ -0,0 +1,172 @@
+        return loss
+
+    def on_fit_start(self):
+        GlobalEnv.current_encoder_epoch = self.current_epoch
+
+    def on_train_epoch_start(self):
+        GlobalEnv.current_encoder_epoch = self.current_epoch
+
+    def training_step(self, batch, batch_idx, optimizer_idx=0):
+        """Training step (pytorch lightning)."""
+        return self._compute_loss(batch, "train")
+
+    def update_metrics(self, batch, batch_idx):
+        target_cat = TileCatalog(self.tile_slen, batch["tile_catalog"])
+
+        sample_tile_cat = self.sample(batch, use_mode=True)
+        sample_cat = sample_tile_cat.to_full_catalog()
+        self.mode_metrics.update(target_cat, sample_cat, None)
+
+        sample_tile_cat = self.sample(batch, use_mode=False)
+        sample_cat = sample_tile_cat.to_full_catalog()
+        self.sample_metrics.update(target_cat, sample_cat, None)
+
+        self.sample_image_renders.update(
+            batch,
+            target_cat,
+            sample_tile_cat,
+            sample_cat,
+            self.current_epoch,
+            batch_idx,
+        )
+
+    def validation_step(self, batch, batch_idx):
+        """Pytorch lightning method."""
+        self._compute_loss(batch, "val")
+        self.update_metrics(batch, batch_idx)
+
+    def report_metrics(self, metrics, logging_name, show_epoch=False):
+        for k, v in metrics.compute().items():
+            self.log(f"{logging_name}/{k}", v, sync_dist=True)
+
+        for metric_name, metric in metrics.items():
+            if hasattr(metric, "plot"):  # noqa: WPS421
+                plot_or_none = metric.plot()
+                name = f"Epoch:{self.current_epoch}" if show_epoch else ""
+                name += f"/{logging_name} {metric_name}"
+                if self.logger and plot_or_none:
+                    fig, _axes = plot_or_none
+                    self.logger.experiment.add_figure(name, fig)
+
+        metrics.reset()
+
+    def on_validation_epoch_end(self):
+        self.report_metrics(self.mode_metrics, "val/mode", show_epoch=True)
+        self.report_metrics(self.sample_metrics, "sample/mode", show_epoch=True)
+        self.report_metrics(self.sample_image_renders, "val/image_renders", show_epoch=True)
+
+    def test_step(self, batch, batch_idx):
+        """Pytorch lightning method."""
+        self._compute_loss(batch, "test")
+        self.update_metrics(batch, batch_idx)
+
+    def on_test_epoch_end(self):
+        self.report_metrics(self.mode_metrics, "test/mode", show_epoch=False)
+        self.report_metrics(self.sample_metrics, "test/mode", show_epoch=False)
+
+    def predict_step(self, batch, batch_idx, dataloader_idx=0):
+        """Pytorch lightning method."""
+        with torch.no_grad():
+            return {
+                "mode_cat": self.sample(batch, use_mode=True),
+                # we may want multiple samples
+                "sample_cat": self.sample(batch, use_mode=False),
+            }
+
+    def configure_optimizers(self):
```

Comment:
Could you omit some of these methods, which are the same in the parent class?

Reply:
resolved